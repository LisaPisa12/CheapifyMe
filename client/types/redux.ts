/* global google */
export type coords = {
  latitude: number;
  longitude: number;
};

export type offer = {
  id: string;
  consumableType: string;
  offerType: string;
  start: string;
  end?: string;
  repeat: boolean;
  repeatEvery?: string;
  description: string;
  score: number;
  available: boolean;
  voted?: boolean;
};
export type location = {
  type: string;
  coordinates: [number, number];
};

export type place = {
  id: string;
  name?: string;
  location?: location;
  address?: string;
  image?: string;
  isInsideRadius?: boolean;
  offers?: offer[];
};

export interface RootState {
  showFloat: boolean;
  scriptLoaded: boolean;
  selectedId: number;
  serviceAPI?: google.maps.places.PlacesService;
  newPlace?: {
    name: string;
    location: {
      latitude: number;
      longitude: number;
    };
  };
  mapCoords: coords;
  userCoords: coords;
  places: place[];
  filteredPlaces: place[];
}

export interface state {
  showFloat: boolean;
  scriptLoaded: boolean;
  selectedId: number;
  serviceAPI?: google.maps.places.PlacesService;
  newPlace?: {
    name: string;
    location: {
      latitude: number;
      longitude: number;
    };
  };
  mapCoords: coords;
  userCoords: coords;
  places: place[];
  filteredPlaces: place[];
}
