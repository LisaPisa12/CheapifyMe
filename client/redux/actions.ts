import { coords, place } from '../types/redux';

export function setScriptLoaded(loaded: boolean) {
  return {
    type: 'SET_SCRIPT_LOADED',
    payload: loaded,
  };
}

export function setMapCoordinates(coords: coords) {
  return {
    type: 'SET_MAP_COORDINATES',
    payload: coords,
  };
}

export function setIfInsideRadius(place: place) {
  return {
    type: 'SET_IF_INSIDE_RADIUS',
    payload: place,
  };
}

export function setUserCoordinates(coords: coords) {
  return {
    type: 'SET_USER_COORDINATES',
    payload: coords,
  };
}

export function setShowFloat(showFloat: boolean) {
  return {
    type: 'SET_SHOW_FLOAT',
    payload: showFloat,
  };
}

export function setPlaces(places: place[]) {
  return {
    type: 'SET_PLACES',
    payload: places,
  };
}
export function setServiceAPI(service: any) {
  return {
    type: 'SET_SERVICE_API',
    payload: service,
  };
}
export function setNewPlace(newPlace: any) {
  return {
    type: 'SET_NEW_PLACE',
    payload: newPlace,
  };
}
export function setNewOffer(place: place) {
  return {
    type: 'SET_NEW_OFFER',
    payload: place,
  };
}

export function setSelectedId(id: number) {
  return {
    type: 'SET_SELECTED_ID',
    payload: id,
  };
}

export function setFilteredPlaces(type: string) {
  return {
    type: 'SET_FILTERED_PLACES',
    payload: type,
  };
}
