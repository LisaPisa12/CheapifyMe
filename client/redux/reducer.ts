import { state } from '../types/redux';
import { createStore, applyMiddleware } from 'redux';
import { createReducer } from '@reduxjs/toolkit';
import { composeWithDevTools } from 'redux-devtools-extension';
import ReduxThunk from 'redux-thunk';

const initialState: state = {
  showFloat: false,
  selectedId: -1,
  scriptLoaded: false,
  serviceAPI: undefined,
  coords: {
    latitude: 41.404278,
    longitude: 2.175098
  },
  places: []
};

const reducer = createReducer(initialState, {
  SET_SCRIPT_LOADED: (state, action) => {
    state.scriptLoaded = action.payload;
  },
  SET_COORDINATES: (state, action) => {
    state.coords = action.payload;
  },
  SET_SHOW_FLOAT: (state, action) => {
    state.showFloat = action.payload;
  },
  SET_SELECTED_ID: (state, action) => {
    state.selectedId = action.payload;
  },
  SET_SERVICE_API: (state, action) => {
    state.serviceAPI = action.payload;
  },
  SET_PLACES: (state, action) => {
    state.places = action.payload;
  },
  SET_NEW_PLACE: (state, action) => {
    state.newPlace = action.payload;
  },
  SET_NEW_OFFER: (state, action) => {
    const place = state.places.find((el) => el.id === action.payload.id);
    if (place) place.offers = action.payload.offers;
    else state.places.push(action.payload);
  }
});

export const store = createStore(
  reducer,
  composeWithDevTools(applyMiddleware(ReduxThunk))
);
