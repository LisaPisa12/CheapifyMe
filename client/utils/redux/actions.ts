import { coords, place } from '../../types/redux';

export function setCoordinates(coords: coords) {
  return {
    type: 'SET_COORDINATES',
    payload: coords
  };
}

export function setLoading(isLoading: boolean) {
  return {
    type: 'SET_LOADING',
    payload: isLoading
  };
}

export function setPlaces(places: place[]) {
  return {
    type: 'SET_PLACES',
    payload: places
  };
}

// export function fetchPlaces(coords: coords) {
//   return function (dispatch: () => {}) {
//     const result = fetch(`someURL`, {
//       headers: {
//         Accept: 'application/json'
//       },
//       body: JSON.stringify(coords)
//     })
//       .then((r) => r.json())
//       .then((response: place[]) => {
//         dispatch(setPlaces(response));
//       });
//     console.log(result);
//   };
// }
