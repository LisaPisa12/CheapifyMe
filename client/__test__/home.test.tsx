import Home from '../pages';

import { store } from '../redux/reducer';
import { Provider } from 'react-redux';
import { cleanup, render, screen } from '@testing-library/react';

const renderComponent = () =>
  render(
    <Provider store={store}>
      <Home></Home>
    </Provider>
  );

describe('App layout', () => {
  afterEach(() => {
    cleanup();
  });

  it('Renders the layout', () => {
    const component = renderComponent();
    expect(component.baseElement).toMatchSnapshot();
  });

  it('Renders the section', () => {
    renderComponent();
    expect(screen.queryByTestId('section')).toBeTruthy();
  });

  it('Renders the childs', () => {
    renderComponent();
    expect(screen.queryAllByTestId('child')).toBeTruthy();
  });

  it('Renders the logo', () => {
    renderComponent();
    expect(screen.queryByTestId('img')).toBeTruthy();
  });

  it('Renders the location textbox', () => {
    renderComponent();
    expect(screen.queryByTestId('location-textbox')).toBeTruthy();
  });

  it('Renders the button with the location icon', () => {
    renderComponent();
    expect(screen.queryByTestId('location-button')).toBeTruthy();
  });

  it('Renders the the map image', () => {
    renderComponent();
    expect(screen.queryByTestId('map-img')).toBeTruthy();
  });
});
