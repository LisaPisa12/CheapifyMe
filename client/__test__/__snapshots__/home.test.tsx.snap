// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App layout Renders the layout 1`] = `
<body>
  <div>
    <section
      class="section"
      data-testid="section"
    >
      <div
        class="childs"
        data-testid="child"
      >
        <img
          class="image"
          data-testid="img"
          src="cheapifyme.gif"
        />
      </div>
      <div
        class="childs"
        data-testid="child"
      >
        <input
          class="input"
          data-testid="location-textbox"
          type="textbox"
        />
        <button
          class="button"
        >
          <img
            class="gps_icon"
            data-testid="location-button"
            src="gps.png"
          />
        </button>
      </div>
      <div
        class="childs"
        data-testid="child"
      >
        <img
          class="map_placeholder"
          data-testid="map-img"
          src="map-placeholder.jpg"
        />
      </div>
    </section>
  </div>
</body>
`;
